lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -63,9 +94,442 @@ def __init__(self, *args, distro: Type[OpenSearchDistribution] = None):
 
         self.opensearch = distro(self, PEER)
         self.opensearch_config = OpenSearchConfig(self.opensearch)
-        self.secrets = SecretStore(self)
-        self.relation_store = Store(self)
+        self.peers_data = RelationDataStore(self, PEER)
+        self.secrets = SecretsDataStore(self, PEER)
         self.tls = OpenSearchTLS(self, TLS_RELATION)
+        self.status = Status(self)
+        self.service_manager = RollingOpsManager(
+            self, relation=SERVICE_MANAGER, callback=self._start_opensearch
+        )
+
+        self.framework.observe(self.on.install, self._on_install)
```

Comment:
My understanding is that the lifecycle events should be called in both VM and K8s charm. With the exception of the `*_pebble_ready` event which is proper to K8s.

I think it would make sense **not** to provide a K8s implementation of the `install` callback as it doesn't require a WaitingStatus in a K8s charm as opposed to the VM's, so I'll put that callback on the charm code instead of the base class.

Reply:
Install hooks aren't so useful in k8s charms because they aren't fired when the container is restarted, such as when you delete a pod or restart a controller. The container gets recreated, but whatever happens in the install hook gets skipped, so it's better to add any install steps to the container in k8s.

Reply:
cool! I believe this refactoring (to create a common code base between VMs and K8s) is something we should do in Kafka too. It is however not yet clear to me how this can be achieved, what can be factored out and what can't. Thanks for double checking this. I will mark these points as resolved